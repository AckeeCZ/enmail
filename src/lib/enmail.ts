export interface Message {
    from?: string;
    to?: string | string[];
    content?: string;
    subject?: string;
}

export type SendMessage<AdapterOptions = any, SendOptions = any> =
    (message: Message, adapter: AdapterSend | Adapter<AdapterOptions, SendOptions>, sendOptions?: SendOptions) => Promise<unknown>;
export type AdapterSend<SendOptions = any> = (message: Message, options?: SendOptions) => Promise<unknown>;

export interface Adapter<AdapterOptions = any, SendOptions = any> {
    options: AdapterOptions;
    mailService: any;
    sender: (sendOptions?: SendOptions) => AdapterSend<SendOptions>;
}
