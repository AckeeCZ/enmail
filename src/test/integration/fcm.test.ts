import enmail from '../../index';

jest.mock('firebase-admin', () => ({
    messaging: (_options?: any) => ({
        sendMulticast: jest.fn(data => data),
    }),
    initializeApp: (_options: any) => {},
}));

describe('Firebase cloud messaging adapter', () => {
    const getFcmAdapter = () => enmail.createAdapter.firebaseCloudMessaging(options);
    const options = {
        projectId: 'projectId',
        serviceAccountId: 'serviceAccountId',
    };
    const adapter = getFcmAdapter();
    const sendOptions = { data: { loginUrl: 'https://ackee.cz' } };
    const body = {
        to: Array(200).fill('registrationToken'),
        content: 'Hello world',
        subject: 'Hello',
    };
    it('Fcm adapter can be created', () => {
        expect(typeof adapter.sender).toEqual('function');
        expect(adapter.options).toStrictEqual(options);
        expect(typeof adapter.mailService.sendMulticast).toEqual('function');
        expect(adapter).toMatchSnapshot();
    });
    it('Sender will be successfully returned', () => {
        const sender = enmail.sender(adapter);
        expect(typeof sender).toEqual('function');
        expect(sender).toEqual(adapter.sender);
    });
    it('The notification can be sent via adapter', async () => {
        const fcmMessage = await adapter.sender()(body, sendOptions) as any;
        expect(Object.keys(fcmMessage).sort()).toEqual(['data', 'tokens', 'notification'].sort());
    });
    it('The notification can be sent via adapter with send options', async () => {
        const fcmMessage = await adapter.sender(sendOptions)(body) as any;
        expect(Object.keys(fcmMessage).sort()).toEqual(['data', 'tokens', 'notification'].sort());
    });
    it('The notification can be sent via adapter without send options', async () => {
        const fcmMessage = await adapter.sender()(body) as any;
        expect(Object.keys(fcmMessage).sort()).toEqual(['tokens', 'notification'].sort());
    });
    it('The notification can be sent via enmail.send', async () => {
        const fcmMessage = await enmail.send(body, adapter) as any;
        expect(Object.keys(fcmMessage).sort()).toEqual(['tokens', 'notification'].sort());
    });
    it('The notification can be sent via enmail.send with sendOptions', async () => {
        const fcmMessage = await enmail.send(body, adapter, sendOptions) as any;
        expect(Object.keys(fcmMessage).sort()).toEqual(['data', 'tokens', 'notification'].sort());
    });
    it('The notification can be sent via sender', async () => {
        const fcmMessage = await enmail.send({ ...body, to: 'test' }, adapter.sender()) as any;
        expect(Object.keys(fcmMessage).sort()).toEqual(['tokens', 'notification'].sort());
    });
    it('The notification can be sent via mailService', async () => {
        const message = {
            notification: {
                title: 'Hello',
                body: 'Hello from enmail!',
            },
            tokens: ['test', 'test2'],
            data: {
                visitUrl: 'https://ackee.cz',
            },
        };
        const fcmMessage = await adapter.mailService.sendMulticast(message) as any;
        expect(fcmMessage).toEqual(message);
    });
});
