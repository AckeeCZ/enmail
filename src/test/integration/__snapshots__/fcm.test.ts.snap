// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Firebase cloud messaging adapter Fcm adapter can be created 1`] = `
Object {
  "mailService": Object {
    "sendMulticast": [MockFunction],
  },
  "options": Object {
    "projectId": "projectId",
    "serviceAccountId": "serviceAccountId",
  },
  "sender": [Function],
}
`;
